### PR TITLE
feat: Implement balanced xcall manager

### PR DESCRIPTION
**Pull request by AntonAndell**

### CODE REVIEW

File: contracts/core-contracts/cw-xcall-manager/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+fn array_eq<T>(a: &[T], b: &[T]) -> bool
+where
+    T: Eq + Hash,
+{
+    let a: HashSet<_> = a.iter().collect();
+    let b: HashSet<_> = b.iter().collect();
+
+    a == b
+}
+
```
Comment (ibrizsabin) [inline contracts/core-contracts/cw-xcall-manager/src/contract.rs:218]: we can directly compare 2 vec<String> rust supports equality operator for vec if the element in vec implements `Eq` trait. I have tested it with a unit test as well so we can just do &a==&b.

File: contracts/core-contracts/cw-xcall-manager/README.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+more on how to run tests and develop code. Or go through the
+[online tutorial](https://docs.cosmwasm.com/) to get a better feel
+of how to develop.
+
+[Publishing](./Publishing.md) contains useful information on how to publish your contract
+to the world, once you are ready to deploy it on a running blockchain. And
+[Importing](./Importing.md) contains information about pulling in other contracts or crates
+that have been published.
+
+Please replace this README file with information about your specific project. You can keep
```
Comment (ibrizsabin) [inline contracts/core-contracts/cw-xcall-manager/README.md:97]: maybe update the readme only to include relevant info to the project

File: contracts/cw-common/src/helpers.rs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    deps: &DepsMut,
+    xcall_manager: Addr,
+    protocols: Option<Vec<String>>,
+) -> Result<(), ContractError> {
+    let query_msg = if let Some(x) = protocols {
+        VerifyProtocols { protocols: x }
+    } else {
+        VerifyProtocols { protocols: vec![] }
+    };
+
```
Comment (ibrizsabin) [inline contracts/cw-common/src/helpers.rs:19]: we can just do `VerifyProtocols{protocols: protocols.unwrap_or_default()}`

Comment (ibrizsabin) [reply contracts/cw-common/src/helpers.rs:19]: or we can just revert if protocols are empty?

**Review by ibrizsabin — CHANGES_REQUESTED** [review]
few refactoring concerns

File: contracts/core-contracts/cw-xcall-manager/README.md (hunk 2 above)
Comment (AntonAndell) [reply contracts/core-contracts/cw-xcall-manager/README.md:97]: Lets remove this for now, we can update the most outer one instead later

File: contracts/core-contracts/cw-xcall-manager/src/contract.rs (hunk 1 above)
Comment (AntonAndell) [reply contracts/core-contracts/cw-xcall-manager/src/contract.rs:218]: This fails if the protocols where not specified in the same order, which is important for the proposed removals

File: contracts/cw-common/src/helpers.rs (hunk 3 above)
Comment (AntonAndell) [reply contracts/cw-common/src/helpers.rs:19]: And default just gives !vec[]? 
We can revert since empty just means default for xCall

Comment (ibrizsabin) [reply contracts/cw-common/src/helpers.rs:19]: ya default will give empty vec.

File: contracts/core-contracts/cw-xcall-manager/src/contract.rs (hunk 1 above)
Comment (ibrizsabin) [reply contracts/core-contracts/cw-xcall-manager/src/contract.rs:218]: oh ya . the arrays need to be identical which is not ideal for our case. ironically i noticed the same issue in solidity where order becomes important but it shouldnt be.

**Review by ibrizsabin — APPROVED** [review]

**Pull request merged**